### PR TITLE
feat: merge NMS charm with webui

### PR DESCRIPTION
**Pull request by patriciareinoso**
# Integration tests will not work without the new rock.

# Description

This PR is the merge of the NMS and the Webui charms:

### Relations in the charm:
- ingress
- common_database
- auth_database
- fiveg_gnb_identity
- fiveg_n4
- logging
- sdcore_config

### Libraries:
- data_interfaces
- loki_push_api
- fiveg_gnb_identity
- fiveg_n4
- sdcore_config 
- ingress

### Charm:
- The only mandatory relations are `common_database` and  `auth_database`
- The charm pushes the webui config file `/nms/config/webuicfg.conf` to the container
- The charm pushes the gNB config using the `/config/v1/inventory/gnb` endpoint.
- The charm pushs the UPF config using the `/config/v1/inventory/upf` endpoint.
- The webui service runs on the port `5000`
- The container is called `nms` 
- The service command is `/bin/webconsole --webuicfg {BASE_CONFIG_PATH}/{CONFIG_FILE_NAME}`
- One environment variables was added: `WEBUI_ENDPOINT`. They are needed after the upstream modifications to serve the swagger.

### Tests
UTs were refactored and separated in files:
- fixtures.py :
- test_charm_status.py
- test_charm_workload_configuration.py
- test_gnb_upf_configuration.py
 
# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: tests/integration/test_integration.py [hunk 1]
```diff
@@ -162,100 +129,86 @@ async def get_traefik_ip_address(ops_test: OpsTest) -> str:
...
+
 def _get_host_from_url(url: str) -> str:
     """Return the host from a URL formatted as http://<host>:<port>/ or as http://<host>/."""
     return url.split("//")[1].split(":")[0].split("/")[0]
 
+
 def ui_is_running(nms_endpoint: str) -> bool:
-    """Return whether the UI is running."""
-    url = f"{nms_endpoint}network-configuration"
+    url = f"{nms_endpoint}config/v1/network-slice"
```
Comment (patriciareinoso) [inline tests/integration/test_integration.py:144]: to change back to f"{nms_endpoint}network-configuration" once the new image is used

File: charmcraft.yaml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 name: sdcore-nms-k8s
 
-summary: A Charmed Operator for SD-Core's NMS.
-description: A Charmed Operator for SD-Core's Network Management System (NMS).
+summary: A Charmed Operator for Aether SD-Core's NMS.
```
Comment (patriciareinoso) [inline charmcraft.yaml:3]: Need to update image use after rock is merged

File: src/webui.py
Comment (gruyaume) [inline src/webui.py:1]: Awesome, thank you

File: src/charm.py [hunk 3]
```diff
@@ -265,42 +317,54 @@ def _get_gnb_name_tac_list(self) -> List[Tuple[str, int]]:
...
             if gnb_name and gnb_tac:
-                gnb_name_tac_list.append((gnb_name, int(gnb_tac)))
+                gnb_name_tac_list.append(GnodeB(name=gnb_name, tac=int(gnb_tac)))
         return gnb_name_tac_list
 
-    def _get_upf_config(self) -> str:
-        """Get the UPF configuration for the NMS in a json.
+    def _sync_gnbs(self, webui_gnbs: List[GnodeB], relation_gnbs: List[GnodeB]) -> None:
+        """Align the gNBs from the `fiveg_gnb_identity`relations with the ones in webui."""
+        relation_gnb_names = {gnb.name for gnb in relation_gnbs}
```
Comment (gruyaume) [inline src/charm.py:325]: Why the need for this dict? GnodeB's are dataclasses and should be comparable directly

File: src/charm.py [hunk 4]
```diff
@@ -90,150 +139,158 @@ def __init__(self, *args):
...
+        if not self.model.relations.get(FIVEG_N4_RELATION_NAME):
+            logger.info("Relation %s not available", FIVEG_N4_RELATION_NAME)
+        webui_upf_config = self._webui.get_upfs()
+        relation_upf_config = self._get_upf_config_from_relations()
+        self._sync_upfs(webui_upfs=webui_upf_config, relation_upfs=relation_upf_config)
 
-        Returns:
-            List[Tuple[str, int]]: List of UPF hostnames and ports.
-        """
+    def _configure_gnbs(self) -> None:
```
Comment (gruyaume) [inline src/charm.py:286]: I don't really understand the current function separation between `configure_gnbs` and `sync_gnbs`. I think we can simplify the code a bit by only having a `sync_gnbs()` (taking 0 arguments), calling it directly from `_configure_sdcore_nms` and have the whole logic in `sync_gnbs`.

File: src/charm.py [hunk 5]
```diff
@@ -90,150 +139,158 @@ def __init__(self, *args):
...
+        if not self._auth_database_resource_is_available():
             return
-        self._configure_pebble()
+        self._setup_webui_endpoint_url()
+        self._configure_gnbs()
+        self._configure_upfs()
+        desired_config_file = self._generate_webui_config_file()
+
+        if config_update_required := self._is_config_update_required(desired_config_file):
+            self._write_file_in_workload(WEBUI_CONFIG_PATH, desired_config_file)
```
Comment (gruyaume) [inline src/charm.py:170]: We could move those 3 in a function to simplify the function here, we'd have something like

```python
if not x:
    return
if not y:
    return
do_a()
do_b()
do_c()
```

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: src/charm.py (hunk 3 above)
Comment (patriciareinoso) [reply src/charm.py]: I was over complexifying the method thinking about the case in which a gnb modifies its tac. 
Now it's much simpler

File: src/charm.py (hunk 5 above)
Comment (patriciareinoso) [reply src/charm.py]: I took this from the webui and it's the same behavior that we have in all the charms. 
I was able to simplify it and get rid of the `restart` boolean.
However, I inverted the order, now we create the pebble plan and then write the configuration file. I think it does not break anything.

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: src/charm.py [hunk 6]
```diff
@@ -265,42 +296,54 @@ def _get_gnb_name_tac_list(self) -> List[Tuple[str, int]]:
...
+            if gnb not in relation_gnbs:
+                self._webui.delete_gnb(gnb.name)
+        for gnb in relation_gnbs:
+            if gnb not in webui_gnbs:
+                self._webui.add_gnb(gnb)
+
+    def _sync_upfs(self) -> None:
+        """Align the UPFs from the `fiveg_n4` relations with the ones in webui."""
+        if not self.model.relations.get(FIVEG_N4_RELATION_NAME):
+            logger.info("Relation %s not available", FIVEG_N4_RELATION_NAME)
```
[the remark this answers is not in the document]
Comment (patriciareinoso) [reply src/charm.py:314]: If we return here, the case: one UPF relation, then relation broken, would not update the webui DB.

File: src/charm.py [hunk 7]
```diff
@@ -265,42 +296,54 @@ def _get_gnb_name_tac_list(self) -> List[Tuple[str, int]]:
...
-        gnb_config = []
-        for gnb_name, gnb_tac in gnb_name_tac_list:
-            gnb_conf_entry = {"name": gnb_name, "tac": str(gnb_tac)}
-            gnb_config.append(gnb_conf_entry)
-
-        return json.dumps(gnb_config, sort_keys=True)
+    def _sync_gnbs(self) -> None:
+        """Align the gNBs from the `fiveg_gnb_identity`relations with the ones in webui."""
+        if not self.model.relations.get(GNB_IDENTITY_RELATION_NAME):
+            logger.info("Relation %s not available", GNB_IDENTITY_RELATION_NAME)
```
[the remark this answers is not in the document]
Comment (patriciareinoso) [reply src/charm.py:301]: same as below for `FIVEG_N4_RELATION_NAME`

File: src/charm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (gruyaume) [reply src/charm.py:314]: Ok thanks, good catch

File: src/charm.py (hunk 5 above)
Comment (patriciareinoso) [reply src/charm.py]: not true, we need the file to be written before creating the pebble layer. need to fix this

Comment (patriciareinoso) [reply src/charm.py]: done

File: charmcraft.yaml [hunk 8]
```diff
@@ -30,14 +30,16 @@ assumes:
 storage:
   config:
     type: filesystem
-    minimum-size: 5M
+    minimum-size: 6M
```
[the remark this answers is not in the document]
Comment (patriciareinoso) [reply charmcraft.yaml]: We used to have a 5M and we are adding the config file from webui. But now that we don't have the UPF/gNB file, we cat set 1M as all other charms.

File: README.md [hunk 9]
```diff
@@ -15,18 +15,19 @@ juju deploy sdcore-nms-k8s --channel=edge
 juju deploy traefik-k8s --trust --config external_hostname=<your hostname> --config routing_mode=subdomain
 juju deploy sdcore-upf-k8s --channel=1.5/edge --trust
 juju deploy mongodb-k8s --trust --channel=6/beta
-juju deploy sdcore-webui-k8s --channel=1.5/edge
 juju deploy sdcore-gnbsim-k8s --trust --channel=1.5/edge
-juju integrate mongodb-k8s sdcore-webui-k8s
+juju deploy grafana-agent-k8s --trust --channel=latest/stable
```
Comment (gatici) [inline README.md:19]: For line 9, don't we need to use `juju deploy sdcore-nms-k8s --channel=1.5/edge` ?

File: src/charm.py [hunk 10]
```diff
@@ -16,68 +18,115 @@
...
+from jinja2 import Environment, FileSystemLoader
+from ops import ActiveStatus, BlockedStatus, CollectStatusEvent, ModelError, WaitingStatus
+from ops.charm import CharmBase, EventBase
 from ops.main import main
 from ops.pebble import Layer
+from webui import GnodeB, Upf, Webui
 
 logger = logging.getLogger(__name__)
 
+BASE_CONFIG_PATH = "/nms/config" # "/nms/config"
```
Comment (gatici) [inline src/charm.py:31]:
```suggestion
BASE_CONFIG_PATH = "/nms/config"  # "/nms/config"
```

File: src/charm.py [hunk 11]
```diff
@@ -16,68 +18,115 @@
...
-WORKLOAD_VERSION_FILE_NAME = "/etc/workload-version"
+SDCORE_CONFIG_RELATION_NAME = "sdcore_config"
+AUTH_DATABASE_NAME = "authentication"
+COMMON_DATABASE_NAME = "free5gc"
 GRPC_PORT = 9876
-SDCORE_CONFIG_RELATION_NAME = "sdcore-config"
+WEBUI_URL_PORT = 5000
 WEBUI_SERVICE_NAME = "webui"
 
+def _get_pod_ip() -> Optional[str]:
```
[the remark this answers is not in the document]
Comment (gatici) [reply src/charm.py:47]: 2 blanks lines are needed.

File: src/webui.py [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+JSON_HEADER = {'Content-Type': 'application/json'}
+
+
+@dataclass
+class GnodeB:
+    """Class to represent a gNB."""
+    name: str
+    tac: int
+
+@dataclass
```
Comment (gatici) [inline src/webui.py:26]:
```suggestion

@dataclass
```

Comment (gatici) [reply src/webui.py:26]: 2 blank lines are required.

File: src/webui.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        try:
+            response.raise_for_status()
+        except requests.HTTPError as e:
+            logger.error("Failed to get resource from inventory: %s", e)
+            return []
+        resources = response.json()
+        logger.info("Got %s from inventory", resources)
+        return resources
+
+    def _add_resource_to_inventory(self, url: str, resource_name: str, data: dict) -> None:
```
Comment (gatici) [inline src/webui.py:86]: This method can be static or just a separate function.

File: src/webui.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        response = requests.post(url, headers=JSON_HEADER, json=data)
+        try:
+            response.raise_for_status()
+        except Exception as e:
+            logger.error("Failed to add %s to webui: %s", resource_name, e)
+            return
+        logger.info("%s added to webui", resource_name)
+
+    def _delete_resource_from_inventory(self, inventory_url: str, resource_name: str) -> None:
```
Comment (gatici) [inline src/webui.py:95]: This method can be static or just a separate function.

File: src/webui.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+
+@dataclass
+class Upf:
+    """Class to represent a UPF."""
+    hostname: str
+    port: int
+
+class Webui:
+    """Handle webui API calls."""
+    def __init__(self, url: str):
```
Comment (gatici) [inline src/webui.py:34]:
```suggestion
    
    def __init__(self, url: str):
```

Comment (gatici) [reply src/webui.py:34]: A blank line is required before _init_ method definition.

File: tests/unit/fixtures.py [hunk 16]
```diff
@@ -0,0 +1,130 @@
...
+FIVEG_N4_RELATION_NAME = "fiveg_n4"
+GNB_IDENTITY_RELATION_NAME = "fiveg_gnb_identity"
+REMOTE_APP_NAME = "some_app"
+SDCORE_CONFIG_RELATION_NAME = "sdcore_config"
+CONTAINER = "nms"
+CONTAINER_CONFIG_FILE_PATH = "nms/config/webuicfg.conf"
+
+class NMSUnitTestFixtures:
+
+    patcher_check_output = patch("charm.check_output")
```
Comment (gatici) [inline tests/unit/fixtures.py:22]:
```suggestion
class NMSUnitTestFixtures:
    patcher_check_output = patch("charm.check_output")
```

Comment (gatici) [reply tests/unit/fixtures.py:22]: No need black line here.

File: tests/unit/fixtures.py [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+    patcher_webui_delete_gnb = patch("webui.Webui.delete_gnb")
+    patcher_webui_get_upfs = patch("webui.Webui.get_upfs")
+    patcher_webui_add_upf = patch("webui.Webui.add_upf")
+    patcher_webui_delete_upf = patch("webui.Webui.delete_upf")
+    patcher_webui_set_url = patch("webui.Webui.set_url")
+
+
+    @pytest.fixture()
+    def setUp(self):
+        self.mock_check_output = NMSUnitTestFixtures.patcher_check_output.start()
```
Comment (gatici) [inline tests/unit/fixtures.py:38]:
```suggestion
    patcher_webui_set_url = patch("webui.Webui.set_url")

    @pytest.fixture()
    def setUp(self):
        self.mock_check_output = NMSUnitTestFixtures.patcher_check_output.start()
```

Comment (gatici) [reply tests/unit/fixtures.py:38]: need to remove extra blank line.

File: tests/unit/fixtures.py [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+        )
+        self.harness.update_relation_data(  # type:ignore
+            relation_id=fiveg_n4_relation_id,
+            app_or_unit=REMOTE_APP_NAME,
+            key_values=key_values,
+        )
+        return fiveg_n4_relation_id
+
+    @pytest.fixture()
+    def empty_webui_inventory(self) -> Generator[None, None, None]:
```
Comment (gatici) [inline tests/unit/fixtures.py:128]:
```suggestion
    def empty_webui_inventory(self) -> None:
```

Comment (gatici) [reply tests/unit/fixtures.py:128]: This does not return anything.

File: tests/unit/test_charm_status.py [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+
+
+class TestCharmStatus(NMSUnitTestFixtures):
+
+    def test_given_unit_is_not_leader_when_collect_status_then_status_is_blocked(self):
+        self.harness.set_leader(is_leader=False)
+
+        self.harness.evaluate_status()
+
+        assert self.harness.model.unit.status == BlockedStatus("Scaling is not implemented for this charm")  # noqa: E501
```
Comment (gatici) [inline tests/unit/test_charm_status.py:22]:
```suggestion
        assert self.harness.model.unit.status == BlockedStatus(
        "Scaling is not implemented for this charm"
        ) 
```

File: tests/unit/test_charm_status.py [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+
+    def test_given_common_database_relation_not_created_when_collect_status_then_status_is_blocked(
+        self, auth_database_relation_id
+    ):
+        self.harness.evaluate_status()
+
+        assert self.harness.model.unit.status == BlockedStatus("Waiting for common_database relation to be created")  # noqa: E501
```
Comment (gatici) [inline tests/unit/test_charm_status.py:29]:
```suggestion
        assert self.harness.model.unit.status == BlockedStatus(
        "Waiting for common_database relation to be created"
        ) 
```

Comment (gatici) [reply tests/unit/test_charm_status.py:29]: Generally speaking, this file can be formatted again, then long lines will  disappear.

File: tests/unit/test_charm_workload_configuration.py [hunk 21]
```diff
@@ -0,0 +1,244 @@
...
+    SDCORE_CONFIG_RELATION_NAME,
+    NMSUnitTestFixtures,
+)
+from ops.model import ModelError
+
+EXPECTED_CONFIG_FILE_PATH = "tests/unit/expected_webui_cfg.yaml"
+POD_IP = "1.2.3.4"
+WEBUI_ENDPOINT = f"{POD_IP}:5000"
+
+def read_file_content(path: str) -> str:
```
Comment (gatici) [inline tests/unit/test_charm_workload_configuration.py:25]:
```suggestion
WEBUI_ENDPOINT = f"{POD_IP}:5000"


def read_file_content(path: str) -> str:
```

Comment (gatici) [reply tests/unit/test_charm_workload_configuration.py:25]: Need 2 blank lines.

File: tests/unit/test_gnb_upf_configuration.py [hunk 22]
```diff
@@ -0,0 +1,339 @@
...
+                id="gnb_name_and_tac_are_empty_strings_in_gNB_config",
+            ),
+            pytest.param(
+                GNB_IDENTITY_RELATION_NAME,
+                {"gnb_name": "something", "some": "key"},
+                id="invalid_key_in_gNB_config",
+            ),
+                        pytest.param(
+                FIVEG_N4_RELATION_NAME,
+                {"upf_hostname": "some.host.name"},
```
Comment (gatici) [inline tests/unit/test_gnb_upf_configuration.py:48]:
```suggestion
            ),
    pytest.param(
                FIVEG_N4_RELATION_NAME,
                {"upf_hostname": "some.host.name"},
```

File: tests/unit/test_gnb_upf_configuration.py [hunk 23]
```diff
@@ -0,0 +1,339 @@
...
+        self.harness.add_storage("config", attach=True)
+        existing_upf = Upf(hostname="some.host.name", port=1234)
+        self.mock_get_upfs.return_value = [existing_upf]
+        self.set_n4_relation_data({"upf_hostname": "some.host.name", "upf_port": "1234"})
+        self.set_n4_relation_data({"upf_hostname": "my_host", "upf_port": "4567"})
+
+        self.harness.container_pebble_ready(CONTAINER)
+
+        expected_upf = existing_upf = Upf(hostname="my_host", port=4567)
+        self.mock_add_upf.assert_called_once_with(expected_upf)
```
Comment (gatici) [inline tests/unit/test_gnb_upf_configuration.py:217]:
```suggestion

        expected_upf = Upf(hostname="my_host", port=4567)
        self.mock_add_upf.assert_called_once_with(expected_upf)
```

Comment (gatici) [reply tests/unit/test_gnb_upf_configuration.py:217]: `existing_upf` is not used.

File: tests/unit/test_gnb_upf_configuration.py [hunk 24]
```diff
@@ -0,0 +1,339 @@
...
+        self.harness.add_storage("config", attach=True)
+        self.set_gnb_identity_relation_data({"gnb_name": "some.gnb.name", "tac": "6789"})
+
+        self.harness.container_pebble_ready(CONTAINER)
+
+        self.mock_delete_gnb.assert_called_once_with("old.gnb.name")
+        expected_gnb = GnodeB(name="some.gnb.name", tac=6789)
+        self.mock_add_gnb.assert_called_once_with(expected_gnb)
+
+
```
Comment (gatici) [inline tests/unit/test_gnb_upf_configuration.py:339]:
```suggestion
        self.mock_add_gnb.assert_called_once_with(expected_gnb)
```

Comment (gatici) [reply tests/unit/test_gnb_upf_configuration.py:339]: At the end of file there extra blank lines which can be removed.

**Review by gatici — COMMENTED** [review]
Generally looks good. 
Some comments are added to indicate formatting and linting issues.

**Review by gruyaume — APPROVED** [review]

**Pull request merged**